drivers/sensor/pmw3360/pmw3360.c

```diff
@@ -0,0 +1,938 @@
+	}
+
+	k_busy_wait(T_BEXIT);
+
+	return 0;
+
+error:
+	LOG_ERR("SPI reg write failed");
+
+	return err;
+}
+
+static int update_cpi(struct pmw3360_data *dev_data, u32_t cpi)
+{
+	/* Set resolution with CPI step of 100 cpi
+	 * 0x00: 100 cpi (minimum cpi)
+	 * 0x01: 200 cpi
+	 * :
+	 * 0x31: 5000 cpi (default cpi)
+	 * :
+	 * 0x77: 12000 cpi (maximum cpi)
+	 */
+
+	if ((cpi > PMW3360_MAX_CPI) || (cpi < PMW3360_MIN_CPI)) {
+		LOG_ERR("CPI value %u out of range", cpi);
+		return -EINVAL;
+	}
+
+	/* Convert CPI to register value */
+	u8_t value = (cpi / 100) - 1;
+
+	LOG_INF("Setting CPI to %u (reg value 0x%x)", cpi, value);
+
+	int err = reg_write(dev_data, PMW3360_REG_CONFIG1, value);
+	if (err) {
+		LOG_ERR("Failed to change CPI");
+	}
+
+	return err;
+}
+
+static int update_downshift_time(struct pmw3360_data *dev_data, u8_t reg_addr,
+				 u32_t time)
+{
+	/* Set downshift time:
+	 * - Run downshift time (from Run to Rest1 mode)
+	 * - Rest 1 downshift time (from Rest1 to Rest2 mode)
+	 * - Rest 2 downshift time (from Rest2 to Rest3 mode)
+	 */
+	u32_t maxtime;
+	u32_t mintime;
+
+	switch (reg_addr) {
+	case PMW3360_REG_RUN_DOWNSHIFT:
+		/*
+		 * Run downshift time = PMW3360_REG_RUN_DOWNSHIFT * 10 ms
+		 */
+		maxtime = 2550;
+		mintime = 10;
+		break;
+
+	case PMW3360_REG_REST1_DOWNSHIFT:
+		/*
+		 * Rest1 downshift time = PMW3360_REG_RUN_DOWNSHIFT
+		 *                        * 320 * Rest1 rate (default 1 ms)
+		 */
+		maxtime = 81600;
+		mintime = 320;
+		break;
+
+	case PMW3360_REG_REST2_DOWNSHIFT:
+		/*
+		 * Rest2 downshift time = PMW3360_REG_REST2_DOWNSHIFT
+		 *                        * 32 * Rest2 rate (default 100 ms)
+		 */
+		maxtime = 816000;
+		mintime = 3200;
+		break;
+
+	default:
+		LOG_ERR("Not supported");
+		return -ENOTSUP;
+	}
+
+	if ((time > maxtime) || (time < mintime)) {
+		LOG_WRN("Downshift time %u out of range", time);
+		return -EINVAL;
+	}
+
+	__ASSERT_NO_MSG(mintime > 0);
+
+	/* Convert time to register value */
+	u8_t value = time / mintime;
+
+	LOG_INF("Set downshift time to %u ms (reg value 0x%x)", time, value);
+
+	int err = reg_write(dev_data, reg_addr, value);
+	if (err) {
+		LOG_ERR("Failed to change downshift time");
+	}
+
+	return err;
+}
+
+static int pmw3360_async_init_fw_load_start(struct pmw3360_data *dev_data)
+{
+	int err = 0;
+
+	/* Read from registers 0x02-0x06 regardless of the motion pin state. */
+	for (u8_t reg = 0x02; (reg <= 0x06) && !err; reg++) {
+		u8_t buf[1];
+		err = reg_read(dev_data, reg, buf);
+	}
+
+	if (err) {
+		LOG_ERR("Cannot read from data registers");
+		return err;
+	}
+
+	/* Write 0 to Rest_En bit of Config2 register to disable Rest mode. */
+	err = reg_write(dev_data, PMW3360_REG_CONFIG2, 0x00);
+	if (err) {
+		LOG_ERR("Cannot disable REST mode");
+		return err;
+	}
+
+	/* Write 0x1D in SROM_enable register to initialize the operation */
+	err = reg_write(dev_data, PMW3360_REG_SROM_ENABLE, 0x1D);
+	if (err) {
+		LOG_ERR("Cannot initialize SROM");
+		return err;
```

Comment:
I think it would be better to use such approach in other functions, too. I mean, to return the error code from the `if` block that checks it, even if it is the last check in the function. Such code is easier to maintain and enhance. Also returning `0` insted of `err` at the end of the function would improve readability a bit.

Reply:
This reduces the number of strings (which can be a pain if you needs debug version run on the SoC with small flash). But I agree with you that this is not pretty and will try to get rid of it.

Reply:
goto removed